### DISCUSSION TITLE
Before/After CollectionDisconnect Events

### DISCUSSION THREAD

**Discussion by aurewill** [mikro-orm/mikro-orm#4320] [General]
**Is your feature request related to a problem? Please describe.**
I have a 1:M relation (EntityA : Collection<EntityB>). When using Collection#remove on a Collection<EntityB>, it's my understanding that EntityB is removed from the Collection and this action is also propagated to EntityB such that EntityB no longer has a reference to EntityA. Because of this, when before/afterDelete is triggered, the Collection#remove that led to the delete event has already disconnected EntityA from EntityB so I no longer have access to the EntityA reference if hooked into EntityB.

**Describe the solution you'd like**
I'm not sure if a Collection disconnect event is the answer, but that would solve my problem.

**Describe alternatives you've considered**
I may be getting ahead of myself as I'm guessing there's a way to get what I need using the flush/transaction hooks, however, because these hooks are entity agnostic, it would be nice to steer clear of these if I'm only concerned about a specific entity.

**Comment by B4nan** [comment]
You can detect this via collection updates, you can get those from the UoW. You can get the collection snapshot to see what it was before the change.

**Comment by aurewill** [comment]
Thanks - I've been trying this with before/on/afterFlush, however every uow getter is returning empty objects/arrays. Do you have any context on this?

**Reply by B4nan** [reply]
You won't get anything in `beforeFlush` (only the persist/remove stacks can have something in them, but no changesets or collection updates), as that happens before things are even computed - `onFlush` or `afterFlush` are the right places. Note that you won't get everything in the collection updates, there will be things from M:N collections, for 1:M this is not a collection update, as the collection is only inverse side, the owner is the M:1 end, and you will see a regular delete/update changeset for those - `delete` in case you enable orphan removal, or `update` (setting the M:1 to `null`) if you don't.

> every uow getter is returning empty objects/arrays

Let's be concrete, feel free to check the ORM tests, they would disagree with this :] If everything would be empty, there would be no queries fired.

**Comment by aurewill** [comment]
```
// Subscriber
@Injectable()
export class Subscriber2 implements EventSubscriber {
  private logger = new Logger(Subscriber2.name);

  constructor(
    private readonly entityManager: EntityManager
  ) {
    entityManager.getEventManager().registerSubscriber(this);
  }

  async onFlush(args: FlushEventArgs): Promise<void> {
    const collUpdates = args.uow.getCollectionUpdates();
    const snapshots = collUpdates.map(cu => cu.getSnapshot());

    this.logger.log("SNAPSHOTS: " + JSON.stringify(snapshots));
  }
}

// Where Collection#remove, persisting, and flushing happen
async delete(entityA: EntityA, entityBsToDelete: Set<EntityB>) {

    if (entityBsToDelete) {
      await entityA.entityBs.init();

      for (const entityB of entityA.entityBs) {
            if (entityBsToDelete.has(entityB)) {
                   entityA.entityBs.remove(entityB);
            }
      }
    }

    await this.em.persistAndFlush(entityA);

    return entityA;
}

// Entity defs
@Entity()
@Index({ properties: "tenant" })
@Index({ properties: "updatedAt" })
export class EntityA extends TenantAwareBaseEntity<EntityA, string, "creatorId"> {
  @ManyToMany({
    entity: () => EntityC,
    pivotEntity: () => EntityB,
    fixedOrder: true,
    fixedOrderColumn: "order",
  })
  entityCs = new Collection<EntityC>(this);

  @OneToMany({
    entity: () => EntityB,
    mappedBy: (entityB) => entityB.entityA,
    orderBy: { order: QueryOrder.ASC },
    orphanRemoval: true,
  })
  entityBs = new Collection<EntityB>(this);
}

@Entity()
@Unique({ properties: ["entityA", "entityC"] })
export class EntityB extends BaseEntity<EntityB> {
  [OptionalProps]:
    | "entityAId"
    | "entityCId";

  @ManyToOne(() => EntityA, { hidden: true, onDelete: "cascade", ref: true })
  entityA!: Ref<EntityA>;

  @ManyToOne(() => EntityC, { hidden: true, onDelete: "cascade", ref: true })
  entityC!: Ref<EntityC>;

  @Property({ persist: false })
  get entityAId() {
    return this.entityA.id;
  }

  @Property({ persist: false })
  get entityCId() {
    return this.entityC.id;
  }
}
```

**Reply by B4nan** [reply]
I can't help with that, I need to see a complete/executable/minimal repro - something I can try and it will be easy to understand, ideally no web servers, just a script.
